[WIP] feat(patient) secondary, tertiary addresses, phone numbers, and emails

sourabbanka22 opened the pull request:
**A Feature that adds the ability to define secondary, tertiary, etc. phone numbers, addresses, and
emails for a patient.**

New Feature #1920 
Fixes #1920 .

Below are some Screenshots of the Added Functionality:
**Before Adding PhoneNumber and Email of a Patient:**
<img width="960" alt="Before" src="https://user-images.githubusercontent.com/58421086/83183777-090f4600-a146-11ea-84d2-9387870a20b1.PNG">

**After Adding PhoneNumber and Email:**
<img width="959" alt="After" src="https://user-images.githubusercontent.com/58421086/83183878-3a881180-a146-11ea-8547-ece4b153705e.PNG">

User needs to click on the **Add Button** to add phone numbers, emails, etc to the respective array.

Currently I have commented out few tests related to Phone Number, Email and Address Models

**Please let me know if there are any Modifications or Changes required.**

sourabbanka22 commented:
@jackcmeyer I have made the design changes that you suggested

**Please let me know if there are any Modifications or Changes required.**

jackcmeyer commented on src/patients/patient-slice.ts:
```diff
@@ -318,6 +351,80 @@ export const addDiagnosis = (
+    await dispatch(updatePatient(patient, onSuccess))
+  } else {
+    newPhoneNumberError.message = 'patient.phoneNumbers.error.unableToAdd'
+    dispatch(addPhoneNumberError(newPhoneNumberError))
```
It looks like there should be some errors being displayed with this error being dispatched, however, I don't see that functionality when using the UI.

jackcmeyer commented on src/patients/patient-slice.ts:
```diff
@@ -318,6 +351,80 @@ export const addDiagnosis = (
   }
 }
 
+function validatePhoneNumber(phoneNumber: PhoneNumber) {
```
I think this function should also check the validity of the phone number. Then, wherever phone numbers are being validated, then this function should also be used. 

Type should also be a required field. 


(this is true for phone number, email, and address.

jackcmeyer commented on src/model/PhoneNumber.ts:
```diff
@@ -0,0 +1,5 @@
+export default interface PhoneNumber {
+  id: string
+  phoneNumber: string
+  type: string
```
There are only a certain number of types allowed for a phone (and address and email). I think the type of this should be an enum, representing the different types. 

We should then use that Enum to populate the dropdowns. That way, our models and dropdowns require a change in one spot to stay in-sync.

jackcmeyer commented on src/index.css:
```diff
@@ -102,4 +102,14 @@ code {
   flex-grow: 1;
+}
+
+.addButton{
+  margin-top: 2em;
+  top: 50%;
+  right: 0px;
+}
+
+.tableButton{
+  float: right;
```
This CSS should go in a `GeneralInformation.css` file since it is only valid for the GeneralInformation component.

jackcmeyer commented on src/patients/GeneralInformation.tsx:
```diff
@@ -209,52 +241,460 @@ const GeneralInformation = (props: Props) => {
+                              ])
+                            }}
+                          >
+                            Add Phone Number
```
text needs to be internationalized.

jackcmeyer commented on src/patients/GeneralInformation.tsx:
```diff
@@ -209,52 +241,460 @@ const GeneralInformation = (props: Props) => {
+                        options={[
+                          { label: 'Home', value: 'Home' },
+                          { label: 'Work', value: 'Work' },
+                          { label: 'Temp', value: 'Temp' },
```
The label for `Temp` should probably be `Temporary`. This is my bad, I didn't catch this when I wrote the issue.

jackcmeyer commented on src/patients/GeneralInformation.tsx:
```diff
@@ -209,52 +241,460 @@ const GeneralInformation = (props: Props) => {
+                              ])
+                            }}
+                          >
+                            Add Email
```
label should be internationalized.

jackcmeyer commented on src/patients/GeneralInformation.tsx:
```diff
@@ -209,52 +241,460 @@ const GeneralInformation = (props: Props) => {
+          <table className="table table-hover">
+            <thead className="thead-light">
+              <tr>
+                <th>Type</th>
+                <th>Address</th>
```
headers should be internationalized.

jackcmeyer commented on src/patients/GeneralInformation.tsx:
```diff
@@ -209,52 +241,460 @@ const GeneralInformation = (props: Props) => {
+                              ])
+                            }}
+                          >
+                            Add Address
```
button should be internationalized.

sourabbanka22 commented:
@tehKapa I am working towards fixing the tests

sourabbanka22 commented:
I fixed some issues concerning internationalization of keys and some design changes, I have also updated most of the tests, I need some help with the following 2 tests in the EditPatient Section:

1. expect(PatientRepository.saveOrUpdate).toHaveBeenCalledWith(patient)
2. expect(store.getActions()[4]).toContainEqual(patientSlice.updatePatientSuccess(patient))

kumikokashii commented on src/patients/GeneralInformation.tsx:
```diff
@@ -209,52 +241,448 @@ const GeneralInformation = (props: Props) => {
+                color="success"
+                icon="add"
+                iconLocation="left"
+                onClick={() => {
```
When a phone number is entered and user does not click on the "Add" button, and user saves the form, the phone number is not saved. Perhaps, any form entry can be considered as user input to be validated?

kumikokashii replied:
Similarly, I wonder if we can have a way to delete a phone number by checking if user input is empty or not at the time of validation? Then, we might not need a delete button.

sourabbanka22 commented:
@kumikokashii please let me know your thoughts on the implemented changes.

kumikokashii commented:
@sourabbanka22 Hello, thank you for the updates. Based on playing around with the [preview](https://hospitalrun-fron-git-fork-sourabbanka22-secondary-tertia-873150.hospitalrun.now.sh/patients/new), it fees like some functionalities could use further updates? I picture something like [this](https://hospitalrun-fronten-git-fork-kumikokashii-multiple-conta-d94570.hospitalrun.now.sh/patients/new) (only the phone numbers are mocked for demo). The sample code for this is [here](https://github.com/HospitalRun/hospitalrun-frontend/pull/2113). Hope this helps! 🙏

sourabbanka22 commented:
@tehKapa I added the Missing tests for Phone Number, Email and Address Model, but it looks like there are still more tests required I am working on adding more of them.

@kumikokashii do I need to add any more updates on the current functionality?

kumikokashii commented:
@sourabbanka22 Hello, thank you for the updates.

So, as you can see in the other fields in GeneralInformation, like "Given Name", "Family Name", etc., each piece of info can be a component instance. For example, the name fields use `TextInputWithLabelFormGroup`.

This is nice because you just provide data and functions as props, and then let the component do all the work for you.

With the phone, email, address fields, you can create a new component and recycle it. This is a [sample code](https://github.com/HospitalRun/hospitalrun-frontend/pull/2113/files#diff-cf2ea4febf217dcd6a894957f1302376). Please note it's just a starting point reference.

Also, it'd be much appreciated if the above review comments could be revisited and resolved where possible :)

sourabbanka22 commented:
Hi @kumikokashii , I have created component instances for Phone Numbers, Emails and Addresses, does the code look good to you?

@tehKapa I have added all the required tests for the Added Functionalities, please let me know if there is anything else that I can help with

kumikokashii commented:
@sourabbanka22 Thank you for the updates 🙏 It's looking almost ready to go. I'm gonna jump in and tweak it a bit. Please stay tuned. Thanks again!

kumikokashii commented:
In addition to allowing more than one phone number, etc, updating the `GeneralInformation` component to have its `onChange` prop to pass the entire `patient` instead of a key and a value. Due to high volume of changes, transitioning to another branch. Please see https://github.com/HospitalRun/hospitalrun-frontend/pull/2113.
